csrc/preseg_passes/exact_mapped_extent_substitution.cpp

```diff
@@ -22,8 +22,55 @@ namespace {
 // otherwise, validateDomainEquivalence fails. If we really want to substitute,
 // we may need to skip or modify validateDomainEquivalence.
 inline bool isNonSubstitutableID(const IterDomain* id) {
-  return (id->isBroadcast() && !id->hasExpandedExtent()) || id->definition() ||
-      id->extent()->definition();
+  return id == nullptr || (id->isBroadcast() && !id->hasExpandedExtent()) ||
+      id->definition() || id->extent()->definition();
+}
+
+// Build disjoint set of extents from  disjoint set of ids
+auto buildExtentSetFromIdSets(const DisjointSets<Val*>& id_sets) {
+  DisjointSets<Val*> extent_sets;
+  // Loop over each id set
+  for (const auto& id_set_ptr : id_sets.disjointSets()) {
+    // If one of the extent in this set is already in the extent_sets, then
+    // map all other extents to the same set, otherwise create a new set.
+    DisjointSets<Val*>::DisjointSet current_extent_set = nullptr;
+
+    // First loop over the set, to check if one of the extent is already mapped
+    for (auto id_set_val : *id_set_ptr) {
+      auto id = dynamic_cast<IterDomain*>(id_set_val);
+      if (isNonSubstitutableID(id)) {
+        continue;
+      }
+      if (extent_sets.mappingExists(id->extent())) {
+        current_extent_set = extent_sets.disjointSetMap().at(id->extent());
+        break;
+      }
+    }
+
+    // Create a new set if no extent is mapped
+    if (current_extent_set == nullptr) {
+      auto id = dynamic_cast<IterDomain*>(*id_set_ptr->begin());
```

Comment:
Shouldn't it check `isNonSubstitutableID`?

Reply:
You are right. Modified to use the first substitutable id which was set in the 1st loop.
```
    // Create a new set if no extent is mapped
    // if no substitutable id in this set, no need to create a new set, break
    // from the loop
    if (current_extent_set == nullptr) {
      if (first_substitutable_id == nullptr) {
        break;
      }
      auto extent = first_substitutable_id->extent();
      auto it = extent_sets.initializeSet(extent).first;
      current_extent_set = it->second;
    }
```

Reply:
Here we should use `continue` instead of `break`